server/services/v1/query_executor.go

```diff
@@ -0,0 +1,237 @@
+// will create a QuerySession and then will execute the query. For explicit transaction, Begin/Commit/Rollback is
+// creating/storing/removing the QuerySession.
+type QuerySessionExecutor struct {
+	sync.RWMutex
+
+	txMgr     *transaction.Manager
+	tenantMgr *metadata.TenantManager
+	versionH  *metadata.MetaVerHandler
+	tracker   *querySessionTracker
+}
+
+func NewQuerySessionExecutor(txMgr *transaction.Manager, tenantMgr *metadata.TenantManager, versionH *metadata.MetaVerHandler) *QuerySessionExecutor {
+	return &QuerySessionExecutor{
+		txMgr:     txMgr,
+		tenantMgr: tenantMgr,
+		versionH:  versionH,
+		tracker:   newQuerySessionTracker(),
+	}
+}
+
+// CreateQuerySession returns the QuerySession after creating all the necessary elements that a query execution needs.
+// It first creates or get a tenant, read the metadata version and based on that reload the tenant cache and then finally
+// create a transaction which will be used to execute all the query in this session.
+func (executor *QuerySessionExecutor) CreateQuerySession(ctx context.Context, reloadVerOutside bool) (*QuerySession, error) {
+	tenant, err := executor.tenantMgr.CreateOrGetTenant(ctx, executor.txMgr, metadata.NewDefaultNamespace())
+	if err != nil {
+		return nil, err
+	}
+
+	var version metadata.Version
+	if reloadVerOutside {
+		version, err = executor.versionH.ReadInOwnTxn(ctx, executor.txMgr)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	tx, err := executor.txMgr.StartTx(ctx)
+	if err != nil {
+		return nil, err
+	}
+	txCtx := tx.GetTxCtx()
+
+	if reloadVerOutside {
+		// use version calculated outside
+		if err = tenant.ReloadUsingOutsideVersion(ctx, tx, version, txCtx.Id); ulog.E(err) {
+			return nil, err
+		}
+	} else {
+		// safe to read version in a transaction
+		if err = tenant.ReloadUsingTxVersion(ctx, tx, txCtx.Id); ulog.E(err) {
+			return nil, err
+		}
+	}
+
+	return &QuerySession{
+		tx:      tx,
+		txCtx:   txCtx,
+		tenant:  tenant,
+		version: version,
+	}, nil
+}
+
+func (executor *QuerySessionExecutor) GetQuerySession(id string) *QuerySession {
+	return executor.tracker.GetSession(id)
+}
+
+func (executor *QuerySessionExecutor) AddQuerySession(id string, session *QuerySession) {
+	executor.tracker.PutSession(id, session)
+}
+
+func (executor *QuerySessionExecutor) RemoveQuerySession(id string) {
+	executor.tracker.RemoveSession(id)
+}
+
+// Execute is responsible to execute a query. In a way this method is managing the lifecycle of a query. For implicit
+// transaction everything is done in this method. For explicit transaction, a session may already exist so it only
+// needs to run without calling Commit/Rollback.
+func (executor *QuerySessionExecutor) Execute(ctx context.Context, req *ReqOptions) (*Response, error) {
+	if req.txCtx != nil {
+		id := req.txCtx.Id
+		session := executor.tracker.GetSession(id)
+		resp, ctx, err := session.Run(ctx, req.queryRunner)
+		session.ctx = ctx
+		return resp, err
+	} else {
+		start := time.Now()
```

Comment:
Should retrying logic go into `executeWithRetry`?

Reply:
it simplifies the defer logic but no preference can be moved.

Reply:
ok. then rename it to something else maybe, cause it's called 'withRetry` while it's not retrying anything.

Reply:
I moved the logic in the method.